rachellegelden/Home-Together-Canada
# HTC-466/467: routes to update member account and profile

# [HTC-366](https://github.com/rachellegelden/Home-Together-Canada/issues/<366>) & [HTC-367](https://github.com/rachellegelden/Home-Together-Canada/issues/<366>) 

## Summary
- Added the routes that will be used to update members' account info and profile. This consists of 4 main parts:
  - update the member's info (this is mainly the values stored in abstract user)
  - update the basic fields of the fields of the member's profile (ie. these are the fields stored in member account table)
  - update the member's status. Depending on the status, this may require updates to the LivesWiths table (where we keep track of roommates)
  - update areas of interest. This requires updates to the areaOfInterest table
- Added the areas of interest and the roommates for a member to the `member/profile` route (I missed this in #409)

## Relevant Motivation & Context
This will be used in the members account summary portion of the site to update their information

## Testing Instructions
### Testing updating member's account info
1. create a user
2. using Postman, send a `POST`  request to `http://localhost:3001/member/info/update/` with the following body:
```
{
    "email": "member1@gelden.com",
    "firstName": "Rachelle",
    "lastName": "Gelden",
    "phoneNumber": 4033365156,
    "addressLine1": "1 Oak Ave",
    "city": "Kelowna",
    "province": "BC",
    "postalCode": "T1V1S4",
    "hasDifferentMailingAddress": false
}
```
Feel free to change the values.
4. Next send a `GET` request to `http://localhost:3001/member/info/` and make sure that the updates to the member's info is as expected

### Testing updating a member's profile
1. send a `POST` request to `http://localhost:3001/member/profile/update/` with the following body:
```
{
    "gender": "Female",
    "birthYear": 1998,
    "status": "Couple",
    "workStatus": "Retired",
    "minMonthlyBudget": 500,
    "maxMonthlyBudget": 950,
    "hasHomeToShare": false,
    "isReligionImportant": false,
    "isDietImportant": false,
    "hasHealthMobilityIssues": false,
    "hasAllergies": false,
    "hasPets": false,
    "isSmoker": false,
    "numRoommates": 1
}
```
Feel free to change the values.
4. Next send a `GET` request to `http://localhost:3001/member/profile/` and make sure that the updates to the member's info is as expected

### Testing updating users status
The basic process for updating the status is as follows:
1. make sure that you are logged in -  send a `POST` request to `http://localhost:3001/user/login/` with the following body:
```
{
    "username": "member2",
    "password": "Password123"
}
```
2. send a `POST` request to `http://localhost:3001/member/profile/status/update/` with the following body:
```
{
    "status": "Existing Group",
    "existingGroupUsernames": ["member4", "member5"]
}
```
Note: if you are changing the status to `Couple` or `Couple With Children`, then `existingGroupUsernames` should be `partnerUsername`. 
3. send a `GET` request to `http://localhost:3001/roommates/` and make sure that you see the expected roommates in the list
4.  Next send a `GET` request to `http://localhost:3001/member/profile/` and make sure that the status is what you expect it to be. 

The following cases must be tested (note that each `X.X` is its own separate case and each must be tested. This will require making several accounts and will probably require dropping the DB).
```
case 2 - was a couple but there was a change
// 2.1 covers status changed from couple to existing group
	// delete entry from LivesWith, add new entry to LivesWiths

// 2.2 was a couple, now a different kind of couple
	// delete entry from LivesWith, add new entry to LivesWith

// 2.3 was a couple, now isn't a couple or an EG
	// delete entry from LivesWith

// 2.4 still is a couple, but the partner's username has changed (this should cover removing the partner's username)
	// delete entry from LivesWith, add new entry to LivesWith
	
// 2.5 still a couple but partner's username has been removed
	// delete entry from LivesWiths

// case 3 - old status was EG, but there has been a change
// 3.1 was an EG now is a couple
	// delete entry from LivesWith, add new entries to LivesWith

// 3.2 was an EG, now is a non-couple status
	// delete entry from LivesWith

// 3.3 still is EG, but the group member usernames have changed (this should cover removing the group members' usernames)
	// delete entries from LivesWith, add new entries to LivesWith
	
// 3.4 still EG but usernames have been removed	
	// delete entries from LivesWiths

// case 4 - was a non EG/couple, but there has been a change
// 4.1 now is an EG
	// add new entries to LivesWith

// 4.2 now is couple
	// add new entry to LivesWith

// 4.3 now is a different non EG/couple status
	// just update the status
```

### Testing updating areas of interested
1. create a member (it is easiest if you drop the DB before starting testing this task, this will keep things less cluttered and easier to read)
2. send a `POST` request to `http://localhost:3001/member/profile/areasOfInterest/update/` with the following body:
```
{
    "areasOfInterest": [{"province": "AB", "city": "High River", "radius": 80}]    
}
```
3. send a `GET` request to `http://localhost:3001/areasOfInterest/` and make sure that areas of interest updated as expected

## Developer checklist prior to opening this pull request:


## Reviewer
